airflow/plugins/operators/rectangle_realtime_data_operator.py

```diff
@@ -0,0 +1,214 @@
+
+        # organize rt files by itpId_urlNumber
+        rt_files = []
+        for r in rt:
+            rt_files.append(fs.ls(r))
+
+        entity_files = [
+            item
+            for sublist in rt_files
+            for item in sublist
+            if self.rt_file_substring in item
+        ]
+
+        feed_files = defaultdict(lambda: [])
+
+        for fname in entity_files:
+            itpId, urlNumber = fname.split("/")[-3:-1]
+            feed_files[(itpId, urlNumber)].append(fname)
+
+        # for some reason the fs.glob command takes up a fair amount of memory here,
+        # and does not seem to free it after the function returns, so we manually clear
+        # its caches (at least the ones I could find)
+        fs.dircache.clear()
+
+        # Now our feed files dict has a key of itpId_urlNumber and a list of files to
+        # parse
+        return feed_files
+
+    def get_google_cloud_filename(self, feed, iso_date):
+        itp_id_url_num = "_".join(map(str, feed))
+        prefix = self.filename_prefix
+        return f"{prefix}_{iso_date}_{itp_id_url_num}.parquet"
+
+    def cast_entities(self, entities):
+        # parse any information provided about fields that need to have dtype changed
+        # e.g. timestamps are strings, and latitude may be inferred as an int.
+        # note that due to a pandas bug, we first convert timestamps to
+        # a float, and then to an integer (hence the separate time_float_cast check)
+        # see: https://stackoverflow.com/a/60024263
+
+        return entities.astype(self.time_float_cast).astype(self.cast)
+
+    def get_entity_details(self, x):
+        """
+        Returns a list of dicts containing entity values to be added to a dataframe.
+        See https://gtfs.org/reference/realtime/v2/ for list of possible
+        entity values by message type.
+        """
+        entity = x.get("entity")
+        parsed_entity_details = []
+        if entity is not None:
+            for e in entity:
+                e_details = {
+                    col: self.pull_value(e, source)
+                    for col, source in self.entity_details.items()
+                }
+                parsed_entity_details.append(e_details)
+        return parsed_entity_details
+
+    def execute(self, context):
+        """Process a RT feed of the given type
+
+        Args:
+            rt_type (string): One of "alerts", "trip_updates", "vehicle_positions"
+            execution_date (date): The execution date being processed
+        """
+        fs = get_fs()
+
+        # get execution_date from context:
+        # https://stackoverflow.com/questions/59982700/airflow-how-can-i-access-execution-date-on-my-custom-operator
+        iso_date = str(context["execution_date"]).split("T")[0]
+
+        # fetch files ----
+        feed_files = self.fetch_bucket_file_names(iso_date)
+
+        # parse feeds ----
+        for feed, files in feed_files.items():
```

Comment:
Unsure how you're tracking progress; https://github.com/tqdm/tqdm is my favorite progress bar library

Reply:
Right now we aren't actually tracking progress, tqdm makes sense, I can look into adding